openmmtools/cache.py

```diff
@@ -387,8 +388,10 @@ def get_context(self, thermodynamic_state, integrator=None):
                 # Only one match.
                 context = self._lru[matching_context_ids[0]]
             else:
-                # Multiple matches, prefer non-default Integrator.
-                for context_id in matching_context_ids:
+                # Multiple matches, prefer the non-default Integrator.
+                # Always pick the least recently used to make two consective
+                # calls retrieving the same integrator.
+                for context_id in reversed(matching_context_ids):
                     if context_id[1] != self._default_integrator_id():
                         context = self._lru[context_id]
                         break
```

Comment:
Somewhat related to the problem,  but actually older

Although you make this "prefer" the non default integrator, won't this never select the default integrator? Since you reject every integrator id matching default, I don't see where the default selection actually comes in. Wouldn't you need a line before the `for` loop starts to select the default integrator and replace it in the loop if the `if` condition is met?